File: apps/api/src/domain/product/product.ts
export interface Category {
  readonly id: string
  readonly name: string
}

export interface Product {
  readonly id: string
  readonly sku: string
  readonly name: string
  readonly description: string | null
  readonly price: number
  readonly stock: number
  readonly weightKg: number | null
  readonly category: Category | null
  readonly createdAt: Date
  readonly updatedAt: Date
}

export function normalizeSku(sku: string): string {
  return sku.trim().toUpperCase()
}

export function normalizeCategoryName(name: string): string {
  return name.trim()
}

File: apps/api/src/ecommerce_api/domain/product.py
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal


@dataclass(frozen=True, slots=True)
class Category:
    id: str
    name: str


@dataclass(frozen=True, slots=True)
class Product:
    id: str
    sku: str
    name: str
    description: str | None
    price: Decimal
    stock: int
    weight_kg: Decimal | None
    category: Category | None
    created_at: datetime
    updated_at: datetime


def normalize_sku(sku: str) -> str:
    return sku.strip().upper()


def normalize_category_name(name: str) -> str:
    return name.strip()
